### PR TITLE
[RFC] assert: keep "where … and …" output in verbose mode

### PR DESCRIPTION
**Pull request by blueyed**
Ref: https://github.com/pytest-dev/pytest/issues/5932
Fixes: https://github.com/pytest-dev/pytest/issues/5192

### CODE REVIEW

**Comment by blueyed** [comment]
This is quite hacky (but might still pass existing tests..).

I think it might be better to replace something like "~%(default)" instead, which would avoid having to pass through the original `repl`.
OTOH it might be useful in custom hooks.

**Comment by blueyed** [comment]
/cc @darrenburns via https://github.com/darrenburns/pytest-clarity (https://github.com/pytest-dev/pytest/issues/5931)
I guess your feedback would be helpful here.. :)

**Comment by darrenburns** [comment]
This would be really cool! I personally prefer this approach over the templating suggestion, but I guess it may break backwards compatibility since the hook interface will change?

File: src/_pytest/hookspec.py [hunk 1]
```diff
@@ -472,7 +472,7 @@ def pytest_unconfigure(config):
 # -------------------------------------------------------------------------
 
 
-def pytest_assertrepr_compare(config, op, left, right):
+def pytest_assertrepr_compare(config, op, left, right, expl):
```
Comment (nicoddemus) [inline src/_pytest/hookspec.py]: Please update the docstring with `expl` as well. 😁

File: src/_pytest/assertion/util.py [hunk 2]
```diff
@@ -146,7 +148,7 @@ def assertrepr_compare(config, op, left, right):
                     type_fn = (isdatacls, isattrs)
                     explanation = _compare_eq_cls(left, right, verbose, type_fn)
                 elif verbose > 0:
-                    explanation = _compare_eq_verbose(left, right)
+                    return expl + "\n~" + "\n~".join(_compare_eq_verbose(left, right))
```
Comment (nicoddemus) [inline src/_pytest/assertion/util.py]: Sorry, I don't get the `"\n~"` part...? 🤔

**Comment by nicoddemus** [comment]
> but I guess it may break backwards compatibility since the hook interface will change?

No, `pluggy` is smart and discards parameters that an implementation doesn't declare. 😉

File: src/_pytest/hookspec.py (hunk 1 above)
Comment (blueyed) [reply src/_pytest/hookspec.py]: Would need more updating in general then, especially for when we go with the return-string-to-skip special postprocessing approach.

File: src/_pytest/assertion/util.py (hunk 2 above)
Comment (blueyed) [reply src/_pytest/assertion/util.py]: It mimics https://github.com/blueyed/pytest/blob/1543a16330a5be10022505429b66608e7ee0ed19/src/_pytest/assertion/__init__.py#L139 (which is skipped for strings).
It's the "mini language" used with https://github.com/blueyed/pytest/blob/1543a16330a5be10022505429b66608e7ee0ed19/src/_pytest/assertion/util.py#L59-L91.

**Comment by blueyed** [comment]
Why do you both prefer this over a special template var (keep in mind that you have to use the "mini template language" here already).
It requires you to be aware of internals more, including that you have to prefix things with "~"/">".
It also clearly is a hack to handle strings in a special way (to skip post-processing).

**Comment by blueyed** [comment]
@nicoddemus 
Could you make up your mind?

(it is a long-standing regression, so it would be great to make some progress here)

**Comment by blueyed** [comment]
Who can we ping about this - being more into the assertion rewriting part of pytest?

**Comment by nicoddemus** [comment]
> Could you make up your mind?

Sorry about the delay, I've been busy with other things.

The problem that I have with the `expl` parameter is that the user can't really do anything with it in practice:

```python
def y():
    return [0, 1]

def test():
    assert y() == [1, 2, 3, 4]
```

For this test failure, the `expl` parameter is:

```
"%(py2)s\n{%(py2)s = %(py0)s()\n} == %(py5)s"
```

So the `expl` parameter of `pytest_assertrepr_compare` is only really usefl for pytest's own internal implementation, `assertrepr_compare`.

I see two options:

1. We have to rethink how `pytest_assertrepr_compare` actually works so users can reuse the variables saved by pytest's rewriting mechanism.
2. We need to somehow pass `expl` from the rewriting mechanism directly to `assertrepr_compare` somehow.

I wouldn't even know where to start with 1), and 2) would probably require some hacking, like setting `expl` to a global. 

At first look 2) is ugly, but we might evaluate it under the light that we already are doing something similar by installing `util._reprcompare` during `pytest_runtest_setup`. What if we set `util._reprcompare_expls` during `_call_reprcompare`?

```python
    if util._reprcompare is not None:
        util._reprcompare_expl = expl
        try:
            custom = util._reprcompare(ops[i], each_obj[i], each_obj[i + 1])
        finally:
            util._reprcompare_expl = None
        if custom is not None:
            return custom
```

Again, while this is ugly, at least it is kept internal and won't introduce a parameter which we might regret later.

What do you think?

**Comment by blueyed** [comment]
> The problem that I have with the `expl` parameter is that the user can't really do anything with it in practice:

Yes.

> What if we set `util._reprcompare_expls` during `_call_reprcompare`?

How does that help plugins with using the default?

**Comment by nicoddemus** [comment]
> How does that help plugins with using the default?

Unless I'm misunderstanding, I thought the patch was about to fix "keep "where … and …" output in verbose mode" in the pytest core, not necessarily expose to plugins necessarily.

In my previous post, `assertrepr_compare` would access the global `util._reprcompare_expl` instead of receiving it by argument... again a hack, but we can avoid exposing a `expl` parameter to users, which would require us to carry this over many releases if we ever find a better way to represent the explanation obtained by the rewriter.

**Comment by blueyed** [comment]
@nicoddemus 
The idea here is to also make this available for plugins / do it "properly".  See previous comment: https://github.com/pytest-dev/pytest/pull/6104

**Comment by nicoddemus** [comment]
> The idea here is to also make this available for plugins / do it "properly". See previous comment: #6104

Is that the right issue number? 🤔

**Comment by blueyed** [comment]
> > The idea here is to also make this available for plugins / do it "properly". See previous comment: #6104
> 
> Is that the right issue number?

No, sorry.  More likely https://github.com/pytest-dev/pytest/issues/5932.

**Comment by JBKahn** [comment]
@blueyed does closing this mean no charge coming?

**Comment by blueyed** [comment]
@JBKahn I've closed all my open PRs for now.  Feel free to pick it up, I might come back to it later myself, but just had to clean up things for myself.

**Comment by JBKahn** [comment]
@nicoddemus any more thoughts on this one? We'd really like to be able to use the explanation in the rewrite

**Pull request closed without merging**